### ISSUE TITLE
Handle webkit2gtk-4.0

### ISSUE THREAD

**Issue by probonopd** [VHSgunzo/sharun#2]
webkit2gtk-4.0 needs some extra handling as described in https://github.com/tauri-apps/tauri/pull/2940.

https://github.com/probonopd/PrusaSlicer/blob/efe3df7c9a9a3bd8c2b7bdfa3ec6bb31bb67ab41/.github/workflows/build.yml#L100

is apparently not the way to do it, as it did not work for me.

**Comment by xplshn** [comment]
@probonopd `sharun` simply executes the `ld.so` with `--library-path` using the libraries on `./shared`, it doesn't do anything `deploy-like` other than setting a few common env variables if certain directories are found in the AppDir.

I suppose this is something that you think should be added to `lib4bin`? But even then, shouldn't it work by default with the `strace` mode?

**Comment by probonopd** [comment]
Right, probably needs to be handled in `lib4bin`, which is also mentioned [here](https://github.com/VHSgunzo/sharun?tab=readme-ov-file#usage-lib4bin). I'm still new to all of this ;-)

**Comment by Samueru-sama** [comment]
Tauri does this to the libs:

```
#binary patch absolute paths in libwebkit files
find "$APPDIR"/usr/lib* -name 'libwebkit*' -exec sed -i -e "s|/usr|././|g" '{}' \;
```

But why? Also doesn't this mean that they need to change the working directory to the AppDir?

What I mentioned [here](https://github.com/VHSgunzo/sharun/issues/1#issuecomment-2492780899) would be a different problem because we are also shipping the dynamic linker, maybe that tauri patch isn't needed and just the binaries need to be wrapped with sharun hopefully.

**Comment by probonopd** [comment]
I assume that libwebkit wants to launch the webkit helper binaries from a compiled-in path in `/usr`. So they are patching that to `././` and probably `cd` to `AppDir/usr` prior to executing the main application binary. As a result, libwebkit might launch the webkit helper binaries from inside the AppDir. Quite a hack but I have done similar things against compiled-in absolute paths in the past.

**Comment by probonopd** [comment]
```
user@DESKTOP-4U17RT8:/mnt/c/Users/User/Downloads/squashfs-root$ ./AppRun

(prusa-slicer:143): Gtk-CRITICAL **: 08:37:06.918: gtk_window_resize: assertion 'width > 0' failed
[2024-11-22 08:37:07.175705] [0x00007faf8b8fb7c0] [error]   UserAccount: Failed to read token - no datafile found.
[2024-11-22 08:37:09.009464] [0x00007faf3cff9640] [error]   DBus Connection Error: /usr/bin/dbus-launch terminated abnormally without any error message
[2024-11-22 08:37:09.009503] [0x00007faf3cff9640] [error]   Dbus Messages listening terminating.
08:37:11 AM: Debug: window wxTreeCtrl@0x55557fed9ea0 ("treeCtrl") lost focus even though it didn't have it
08:37:11 AM: Debug: window wxTreeCtrl@0x55557fed9ea0 ("treeCtrl") lost focus even though it didn't have it

** (prusa-slicer:143): ERROR **: 08:37:13.934: Unable to spawn a new child process: Failed to spawn child process “/usr/lib/x86_64-linux-gnu/webkit2gtk-4.0/WebKitNetworkProcess” (No such file or directory)
Trace/breakpoint trap (core dumped)
```

It is looking for `WebKitNetworkProcess` at a compiled-in absolute path.

Maybe we need to resort to `././` patching if we find no other way?

**Comment by Samueru-sama** [comment]
>Maybe we need to resort to ././ patching if we find no other way?

Most likely, **also sharun the binaries in the `webkit2gtk-4.0` dir** which is basically just moving the binaries to the `./bin` dir and putting a relative symlink with their original place and name that points to sharun. 

You can also set `SHARUN_WORKING_DIR=${SHARUN_DIR}` on the `.env` file to make sharun cd into the AppDir when running (it is the equivalent of `cd "$HERE"` on the typical shell `AppRun`) if you go with the patching route.

**Comment by Samueru-sama** [comment]
Getting closer: 

![image](https://github.com/user-attachments/assets/ae00fe86-3489-4247-96f4-1dc60b090751)

Now I no longer crash when I click on printables which used to give the webkitgtk error. 

I did the following to fix the issue: 

Make a `.env` file in the top level of the AppDir with this: 

```
SHARUN_WORKING_DIR=${SHARUN_DIR}
LIBGL_DRIVERS_PATH=${SHARUN_DIR}/shared/lib/dri
```

```
mv ./shared/lib/webkit2gtk-4.0/WebKitWebProcess ./shared/bin
mv ./shared/lib/webkit2gtk-4.0/WebKitNetworkProcess ./shared/bin
mv ./shared/lib/webkit2gtk-4.0/MiniBrowser ./shared/bin

( cd ./shared/lib/webkit2gtk-4.0
  ln -s ../../../sharun ./WebKitWebProcess
  ln -s ../../../sharun ./WebKitNetworkProcess
  ln -s ../../../sharun ./MiniBrowser
)

find ./shared/lib -name 'libwebkit*' -exec sed -i 's|/usr|././|g' {} \;

ln -s ./ ./shared/lib/x86_64-linux-gnu
```

**Comment by Samueru-sama** [comment]
Relevant issue about the EGL_BAD_PARAMETER error: https://github.com/gitbutlerapp/gitbutler/issues/5282

Seems like an older versio nof webkit2gtk has to be used? this thing is a nightmare lol.

**Comment by Samueru-sama** [comment]
Update I was able to reduce most of the strace missing libs: I had to copy the following libs from ubuntu into the AppDir: 

```
libGLX_indirect.so.0
libGLX_mesa.so.0
libxcb-dri2.so.0
libxcb-dri3.so.0
libxcb-glx.so.0
libxcb-present.so.0
libxcb-sync.so.1
libxcb-xfixes.so.0
libxshmfence.so.1
libXxf86vm.so.1
```

Now I only get `"/home/runner/work/PrusaSlicer/PrusaSlicer/deps/build/destdir/usr/local/lib/wx/3.2/web-extensions"` missing when I click on Printables.

**Comment by xplshn** [comment]
![image](https://github.com/user-attachments/assets/e99088a6-0582-448d-ac71-88ea65e0b516)
I made a working one using: `pelfCreator  --maintainer "xplshn" --name "prusaSlicer" --pkg-add "prusa-slicer" --entrypoint "PrusaSlicer.desktop"`

I should check if I can make it smaller via using `pelfCreator`'s `--sharun` mode + `--keep` to keep some /usr, /opt, /var, whatever, files.

![image](https://github.com/user-attachments/assets/1955b63f-88b3-47b3-866b-586e0a5af662)
I used pelfCreator to create one based on Devuan, and the other one which is based on a CachyOS rootfs, you could probably just roll with this after passing the `--getrid` flag and indicating some files & dirs which could be removed from the appdir.

**Comment by probonopd** [comment]
Regarding EGL: Do we also need the `__EGL_VENDOR_LIBRARY_DIRS `environment variable?

**Comment by probonopd** [comment]
@xplshn: Interesting as well. How far can you bring the size down in you approach without impacting functionality?

**Comment by probonopd** [comment]
As of https://github.com/VHSgunzo/sharun/issues/1#issuecomment-2495508221 now, the remaining issue on Fedora is that it claims to need glib-networking for TLS/SSL to work.

![image](https://github.com/user-attachments/assets/3ff75465-7c8e-4040-9b75-4855bbbf7e59)

This message seems to come from libsoup-2.4.so.1.11.2.

Apart from WebKit not working everything else seems to run.

**Comment by probonopd** [comment]
On openSUSE Tumbleweed 20231208:

![image](https://github.com/user-attachments/assets/8315a476-6763-4b15-b0e7-7c7d9c12c600)

**Comment by xplshn** [comment]
> @xplshn: Interesting as well. How far can you bring the size down in you approach without impacting functionality?
> 
> Maybe you could use something like https://github.com/jacereda/fsatrace to find out which files are actually needed and should be kept?

Well, depends the mode. The smallest AppBundle that uses an entire rootfs is 10 megs, 6 out of those 10 megs are the static tools being included...

The `sharun` mode would run `lib4bin` on the binaries which you pass an ditch the rootfs, only keeping the directories you want via `--keep`, also, AppBundles generated like this will always use the directories within ./appdir/proto. So, if you had, ./appdir/proto/usr/share/myapp, the program would be able to locate its config, it'd believe that its config is sitting at /usr/share/myapp. This approach works for all directories, not just `usr`, arbitrary ones too.

**Comment by Samueru-sama** [comment]
>It's a shame that Chromium for Gtk draws in dependencies for GNOME such as https://wiki.gnome.org/Projects/libsoup which in turn draws in things like glib-networking which in turn seems to draw in GIO...
>
>Apparently we need to deploy that stuff (and its dependencies) too, and set GIO_MODULE_DIR, as I have done in my latest [build](https://github.com/probonopd/PrusaSlicer/actions/runs/11988611506?notification_referrer_id=NT_kwDOACXZubMxMzUxNTkwNDIwMDoyNDgwNTY5).

sharun should already set `GIO_MODULE_DIR` [automatically](https://github.com/VHSgunzo/sharun/commit/558fddda206d3ab2d05862dab3476f9e7bb5572a).

**Comment by probonopd** [comment]
So, for my personal test case I found solutions for all of the above. But it is a lot of manual work, Is it in the scope of this project to provide automation for this? Thanks.

**Comment by Samueru-sama** [comment]
> So, for my personal test case I found solutions for all of the above. But it is a lot of manual work, Is it in the scope of this project to provide automation for this? Thanks.

This is a very specific issue with webkit2gtk. 

With that said I was recently shared [this amazing thing](https://github.com/fritzw/ld-preload-open) that can be used to overwrite paths. 

TLDR: 

```
git clone https://github.com/fritzw/ld-preload-open.git
cd ld-preload-open
make all
export PATH_MAPPING="/usr/lib:/$HERE/shared/lib"
LD_PRELOAD="$HERE/path-mapping.so" "$HERE"/AppRun
```

And now calls to `/usr/lib` will be redirected to `$HERE/shared/lib` in this example. 

So maybe `lib4bin` can have something that builds this library and preloads it when it finds in the AppDir.

**Comment by probonopd** [comment]
Yes... but it needs to be considered also:
* Not all binaries are dynamic
* Exporting `LD_PRELOAD` means that not only the main process started by AppRun, but also all subprocesses started by it might see it, which might not be desirable

**Comment by Samueru-sama** [comment]
>Not all binaries are dynamic

In that case the binary works already. I don't know of a scenario where we ship a static binary in an AppImage that is hardcoded to look into `/usr/lib`?

>Exporting LD_PRELOAD means that not only the main process started by AppRun, but also all subprocesses started by it might see it, which might not be desirable

I guess you could add it with `patchelf --add-needed` as well but I'm not sure if it only needs to be added to all binaries or also all libraries as well though.

**Comment by Samueru-sama** [comment]
Btw I was reading the manpage of [ld.so](https://man7.org/linux/man-pages/man8/ld.so.8.html) and noticed that you can preload libraries with the `--preload` flag which does the same things as `LD_PRELOAD` but has the benefit that it does not propagate to child processes. 

So yeah using [ld-preload-open](https://github.com/fritzw/ld-preload-open) is the way to go in this case.

**Comment by Samueru-sama** [comment]
I made use of the library [here](https://github.com/pkgforge-dev/Dolphin-emu-AppImage/blob/Artixlinux/dolphin-emu-appimage.sh#L74-L80) to fix an issue with the dolphin emulator being hardcoded to look for some files in `/usr/share/dolphin-emu/sys`
